File: LivrusVirtus/BoaConstructor/wxPython/CadEnderecos.py
#Boa:FramePanel:frmCadEnderecos

from wxPython.wx import *
from TextValidator import TextValidator
from Cadastros import frmCadastros
from RotinasGlobais import RotinasGlobais
from ConsultasSQL import ConsultasSQL
from resource import Resource

[wxID_FRMCADENDERECOS, wxID_FRMCADENDERECOSCMBESTADO, 
 wxID_FRMCADENDERECOSEDTBAIRRO, wxID_FRMCADENDERECOSEDTCEP, 
 wxID_FRMCADENDERECOSEDTCIDADE, wxID_FRMCADENDERECOSLABBAIRRO, 
 wxID_FRMCADENDERECOSLABCEP, wxID_FRMCADENDERECOSLABCIDADE, 
 wxID_FRMCADENDERECOSLABESTADO, 
] = map(lambda _init_ctrls: wxNewId(), range(9))

def create(parent):
    return frmCadEnderecos(parent)

class frmCadEnderecos(frmCadastros):
    def _init_utils(self):
        pass

    def _init_ctrls(self, prnt):
        frmCadastros._init_ctrls(self, prnt, TextValidator.FILTER_NUMERIC, False)
        self._init_utils()
        self.SetClientSize(wxSize(523, 171))
        self.labDescricao.SetLabel('Logradouro: ')
        self.labDescricao.SetSize(wxSize(57, 13))
        self.edtDescricao.SetSize(wxSize(264, 21))

        self.labBairro = wxStaticText(id=wxID_FRMCADENDERECOSLABBAIRRO,
              label=u'Bairro:', name=u'labBairro', parent=self, pos=wxPoint(280,
              48), size=wxSize(48, 13), style=0)

        self.edtBairro = wxTextCtrl(id=wxID_FRMCADENDERECOSEDTBAIRRO,
              name=u'edtBairro', parent=self, pos=wxPoint(280, 64),
              size=wxSize(144, 21), style=0, value=u'')
        self.edtBairro.Enable(False)

        self.labCEP = wxStaticText(id=wxID_FRMCADENDERECOSLABCEP,
              label=u'CEP: ', name=u'labCEP', parent=self, pos=wxPoint(8, 88),
              size=wxSize(40, 13), style=0)

        self.edtCEP = wxTextCtrl(id=wxID_FRMCADENDERECOSEDTCEP, name=u'edtCEP',
              parent=self, pos=wxPoint(8, 104), size=wxSize(72, 21), style=0,
              value=u'')
        self.edtCEP.Enable(False)

        self.labCidade = wxStaticText(id=wxID_FRMCADENDERECOSLABCIDADE,
              label=u'Cidade: ', name=u'labCidade', parent=self, pos=wxPoint(88,
              88), size=wxSize(39, 13), style=0)

        self.edtCidade = wxTextCtrl(id=wxID_FRMCADENDERECOSEDTCIDADE,
              name=u'edtCidade', parent=self, pos=wxPoint(88, 104),
              size=wxSize(184, 21), style=0, value=u'')
        self.edtCidade.Enable(False)

        self.labEstado = wxStaticText(id=wxID_FRMCADENDERECOSLABESTADO,
              label=u'Estado: ', name=u'labEstado', parent=self,
              pos=wxPoint(280, 88), size=wxSize(56, 13), style=0)

        self.cmbEstado = wxChoice(choices=[], id=wxID_FRMCADENDERECOSCMBESTADO,
              name=u'cmbEstado', parent=self, pos=wxPoint(280, 104),
              size=wxSize(144, 21), style=0, validator=wxDefaultValidator)
        self.cmbEstado.Enable(False)

        self.IniciaBotoes()
      
        self.btnPesquisar.SetDimensions(440, 8, 75, 25)
        self.btnAnterior.SetDimensions(440, 40, 75, 25)
        self.btnProximo.SetDimensions(440, 72, 75, 25)
        self.btnSalvar.SetDimensions(200, 136, 75, 25)
        self.btnExcluir.SetDimensions(280, 136, 75, 25)
        self.btnLimpar.SetDimensions(360, 136, 75, 25)
        self.btnFechar.SetDimensions(440, 136, 75, 25)
      
        self.FazLayout()
      
        self.edtDescricao.SetMaxLength(40)
        self.edtBairro.SetMaxLength(15)
        self.edtCEP.SetMaxLength(8)
        self.edtCidade.SetMaxLength(30)
        self.AdicionaCombo()

        EVT_BUTTON(self.btnNovo, self.btnNovo.GetId(), self.OnBtnNovoClick)
        EVT_BUTTON(self.btnFechar, self.btnFechar.GetId(), 
          self.OnBtnFecharClick)
        EVT_BUTTON(self.btnPesquisar, self.btnPesquisar.GetId(),
              self.OnBtnPesquisarClick)
        EVT_BUTTON(self.btnAnterior, self.btnAnterior.GetId(),
              self.OnBtnAnteriorClick)
        EVT_BUTTON(self.btnProximo, self.btnProximo.GetId(),
              self.OnBtnProximoClick)
        EVT_BUTTON(self.btnSalvar, self.btnSalvar.GetId(),
              self.OnBtnSalvarClick)
        EVT_BUTTON(self.btnExcluir, self.btnExcluir.GetId(),
              self.OnBtnExcluirClick)
        EVT_BUTTON(self.btnLimpar, self.btnLimpar.GetId(),
              self.OnBtnLimparClick)

    def __init__(self, parent):
        self._init_ctrls(parent)

    def FazLayout(self):
      self.sizer_1 = wxBoxSizer(wxVERTICAL)
      self.sizer_2 = wxBoxSizer(wxHORIZONTAL)
      self.sizer_3 = wxBoxSizer(wxHORIZONTAL)
      self.sizer_4 = wxBoxSizer(wxVERTICAL)
      self.sizer_5 = wxBoxSizer(wxVERTICAL)
      self.sizer_6 = wxBoxSizer(wxHORIZONTAL)
      self.grid_sizer_6 = wxGridSizer(2, 1, 0, 0)
      self.grid_sizer_5 = wxGridSizer(2, 1, 0, 0)
      self.grid_sizer_4 = wxGridSizer(2, 1, 0, 0)
      self.sizer_8 = wxBoxSizer(wxHORIZONTAL)
      self.grid_sizer_9 = wxGridSizer(2, 1, 0, 0)
      self.grid_sizer_8 = wxGridSizer(2, 1, 0, 0)
      self.sizer_7 = wxBoxSizer(wxHORIZONTAL)
      self.grid_sizer_7 = wxGridSizer(2, 1, 0, 0)
      self.grid_sizer_7.Add(self.labCodigo, 0, wxALIGN_BOTTOM, 0)
      self.grid_sizer_7.Add(self.edtCodigo, 0, 0, 0)
      self.sizer_7.Add(self.grid_sizer_7, 0, 0, 0)
      self.sizer_7.Add(self.btnNovo, 0, wxLEFT|wxALIGN_BOTTOM, 5)
      self.sizer_5.Add(self.sizer_7, 1, 0, 0)
      self.grid_sizer_8.Add(self.labDescricao, 0, wxALIGN_BOTTOM, 0)
      self.grid_sizer_8.Add(self.edtDescricao, 0, 0, 0)
      self.sizer_8.Add(self.grid_sizer_8, 0, 0, 0)
      self.grid_sizer_9.Add(self.labBairro, 0, wxALIGN_BOTTOM, 0)
      self.grid_sizer_9.Add(self.edtBairro, 0, 0, 0)
      self.sizer_8.Add(self.grid_sizer_9, 0, wxLEFT, 2)
      self.sizer_5.Add(self.sizer_8, 1, wxEXPAND, 0)
      self.grid_sizer_4.Add(self.labCEP, 0, wxALIGN_BOTTOM, 0)
      self.grid_sizer_4.Add(self.edtCEP, 0, 0, 0)
      self.sizer_6.Add(self.grid_sizer_4, 0, 0, 0)
      self.grid_sizer_5.Add(self.labCidade, 0, wxALIGN_BOTTOM, 0)
      self.grid_sizer_5.Add(self.edtCidade, 0, 0, 0)
      self.sizer_6.Add(self.grid_sizer_5, 0, wxLEFT, 4)
      self.grid_sizer_6.Add(self.labEstado, 0, wxALIGN_BOTTOM, 0)
      self.grid_sizer_6.Add(self.cmbEstado, 0, 0, 0)
      self.sizer_6.Add(self.grid_sizer_6, 0, wxLEFT, 5)
      self.sizer_5.Add(self.sizer_6, 0, 0, 0)
      self.sizer_3.Add(self.sizer_5, 0, 0, 0)
      self.sizer_4.Add(self.btnPesquisar, 0, wxALL|wxALIGN_RIGHT, 2)
      self.sizer_4.Add(self.btnAnterior, 0, wxALL|wxALIGN_RIGHT, 2)
      self.sizer_4.Add(self.btnProximo, 0, wxALL|wxALIGN_RIGHT, 2)
      self.sizer_3.Add(self.sizer_4, 1, 0, 0)
      self.sizer_1.Add(self.sizer_3, 1, wxALL|wxEXPAND, 4)
      self.sizer_2.Add(self.btnSalvar, 0, 
        wxALL|wxALIGN_RIGHT|wxALIGN_BOTTOM, 2)
      self.sizer_2.Add(self.btnExcluir, 0, 
        wxALL|wxALIGN_RIGHT|wxALIGN_BOTTOM, 2)
      self.sizer_2.Add(self.btnLimpar, 0, 
        wxALL|wxALIGN_RIGHT|wxALIGN_BOTTOM, 2)
      self.sizer_2.Add(self.btnFechar, 0, 
        wxALL|wxALIGN_RIGHT|wxALIGN_BOTTOM, 2)
      self.sizer_1.Add(self.sizer_2, 0, 
        wxRIGHT|wxBOTTOM|wxALIGN_RIGHT|wxALIGN_BOTTOM, 4)
      self.SetAutoLayout(true)
      self.SetSizer(self.sizer_1)
      self.sizer_1.Fit(self)
      self.sizer_1.SetSizeHints(self)

    def ValidaDados(self, bTodosDados):
      if RotinasGlobais.Rotinas.ValidaCampo(
        self.edtCodigo.GetValue(), Resource.STR_CODNINF):
        return False
      if bTodosDados:
    	  if RotinasGlobais.Rotinas.ValidaCampo(
          self.edtDescricao.GetValue(), Resource.STR_ENDNINF):
      		return False
    	  if RotinasGlobais.Rotinas.ValidaCampo(
          self.edtBairro.GetValue(), Resource.STR_BAIRRONINF):
      		return False
    	  if RotinasGlobais.Rotinas.ValidaCampo(
          self.edtCEP.GetValue(), Resource.STR_CEPNINF):
      		return False
    	  if RotinasGlobais.Rotinas.ValidaCampo(
          self.edtCidade.GetValue(), Resource.STR_CIDADENINF):
      		return False
      return True

    def OnBtnFecharClick(self, event):
      self.GetParent().Close(True)
    
    def OnBtnNovoClick(self, event):
      self.btnNovoClick()
      self.InformaLimpaDadosLocal(False)
      self.HabilitaDadosLocal(True)
      self.NovoDado(ConsultasSQL.ConsSQL.Endereco('N', 
        self.edtCodigo.GetValue(),"","","","",""))
      event.Skip()

    def OnBtnPesquisarClick(self, event):
      if self.PesquisarDados(ConsultasSQL.ConsSQL.Endereco('S', 
        self.edtCodigo.GetValue(),"","","","",""), Resource.STR_ENDNENC):
        self.InformaLimpaDadosLocal(True)
        self.HabilitaDadosLocal(True)
      event.Skip()

    def OnBtnAnteriorClick(self, event):
      if self.registroAnterior():
        self.InformaLimpaDadosLocal(True)
      event.Skip()

    def OnBtnProximoClick(self, event):
      if self.registroProximo():
        self.InformaLimpaDadosLocal(True)
      event.Skip()

    def OnBtnSalvarClick(self, event):
      if self.SalvarDados(self.ValidaDados(True),
        ConsultasSQL.ConsSQL.Endereco('S', self.edtCodigo.GetValue(),
        "","","","",""),
        ConsultasSQL.ConsSQL.Endereco('U', self.edtCodigo.GetValue(), 
        self.edtDescricao.GetValue(),self.edtBairro.GetValue(),
        self.edtCEP.GetValue(),self.edtCidade.GetValue(), 
        self.cmbEstado.GetStringSelection()),
        ConsultasSQL.ConsSQL.Endereco('I', self.edtCodigo.GetValue(), 
        self.edtDescricao.GetValue(),self.edtBairro.GetValue(),
        self.edtCEP.GetValue(),self.edtCidade.GetValue(), 
        self.cmbEstado.GetStringSelection()), True):
        self.InformaLimpaDadosLocal(False)
        self.HabilitaDadosLocal(False)
      event.Skip()

    def OnBtnExcluirClick(self, event):
      if self.ExcluirDados(self.ValidaDados(False),
        ConsultasSQL.ConsSQL.Endereco('D', self.edtCodigo.GetValue(),
        "","","","","")):
        self.InformaLimpaDadosLocal(False)
        self.HabilitaDadosLocal(False)
      event.Skip()

    def OnBtnLimparClick(self, event):
      self.btnLimparClick()
      self.InformaLimpaDadosLocal(False)
      self.HabilitaDadosLocal(False)
      event.Skip()

    def InformaLimpaDadosLocal(self, bInformar):
      if bInformar:
        self.edtBairro.SetValue(self.lstRegistros.col[self.nLinhaRegistro][2])
        self.edtCEP.SetValue(self.lstRegistros.col[self.nLinhaRegistro][3])
        self.edtCidade.SetValue(self.lstRegistros.col[self.nLinhaRegistro][4])
        self.cmbEstado.SetStringSelection(
          self.lstRegistros.col[self.nLinhaRegistro][5])
      else:
        self.edtBairro.Clear()
        self.edtCEP.Clear()
        self.edtCidade.Clear()
        self.cmbEstado.SetSelection(0)
    
    def HabilitaDadosLocal(self, bHabilita):
      self.edtBairro.Enable(bHabilita)
      self.edtCEP.Enable(bHabilita)  
      self.edtCidade.Enable(bHabilita)
      self.cmbEstado.Enable(bHabilita)  

    def AdicionaCombo(self):
      self.cmbEstado.Append(Resource.STR_PE)
      self.cmbEstado.Append(Resource.STR_AC)
      self.cmbEstado.Append(Resource.STR_AL)
      self.cmbEstado.Append(Resource.STR_AP)
      self.cmbEstado.Append(Resource.STR_AM)
      self.cmbEstado.Append(Resource.STR_BA)
      self.cmbEstado.Append(Resource.STR_BR)
      self.cmbEstado.Append(Resource.STR_CE)
      self.cmbEstado.Append(Resource.STR_DF)
      self.cmbEstado.Append(Resource.STR_ES)
      self.cmbEstado.Append(Resource.STR_GO)
      self.cmbEstado.Append(Resource.STR_MT)
      self.cmbEstado.Append(Resource.STR_MS)
      self.cmbEstado.Append(Resource.STR_MG)
      self.cmbEstado.Append(Resource.STR_PA)
      self.cmbEstado.Append(Resource.STR_PB)
      self.cmbEstado.Append(Resource.STR_PR)
      self.cmbEstado.Append(Resource.STR_SC)
      self.cmbEstado.Append(Resource.STR_RN)
      self.cmbEstado.Append(Resource.STR_RS)
      self.cmbEstado.Append(Resource.STR_RJ)
      self.cmbEstado.Append(Resource.STR_RO)
      self.cmbEstado.Append(Resource.STR_RR)
      self.cmbEstado.Append(Resource.STR_SP)
      self.cmbEstado.Append(Resource.STR_SE)
      self.cmbEstado.Append(Resource.STR_TO)

File: LivrusVirtus/wxWindows/wx.NET/VCPP.NET/CadEnderecos.cpp
#include "stdafx.h"
#include "CadEnderecos.h"
#include "Resource.h"
#include "ConsultasSQL.h"
#include "RotinasGlobais.h"

namespace Livrus
{
  frmCadEnderecos::frmCadEnderecos( wx::Window* parent )
  	: frmCadastros(parent, TextValidatorCtrl::FILTER_NUMERIC, false){

    ClientSize = System::Drawing::Size(523, 171);
    labDescricao->Label = "Logradouro:";
    labDescricao->Size = System::Drawing::Size(57, 13);
    edtDescricao->Size = System::Drawing::Size(264, 21);
  
  	labBairro = new wx::StaticText(this, (int)idx::ID_FRMCADENDERECOS_LABBAIRRO, 
      "Bairro:", System::Drawing::Point(280, 48), System::Drawing::Size(48, 13) );
  	edtBairro = new wx::TextCtrl(this, (int)idx::ID_FRMCADENDERECOS_EDTBAIRRO,
      "", System::Drawing::Point(280, 64), System::Drawing::Size(144, 21) );
    edtBairro->Enabled = false;
  	labCEP = new wx::StaticText(this, (int)idx::ID_FRMCADENDERECOS_LABCEP, 
      "CEP: ", System::Drawing::Point(8, 88), System::Drawing::Size(40, 13) );
  	edtCEP = new wx::TextCtrl(this, (int)idx::ID_FRMCADENDERECOS_EDTCEP, 
      "", System::Drawing::Point(8, 104), System::Drawing::Size(72, 21) );
    edtCEP->Enabled = false;
  	labCidade = new wx::StaticText(this, (int)idx::ID_FRMCADENDERECOS_LABCIDADE, 
      "Cidade: ", System::Drawing::Point(88, 88), System::Drawing::Size(39, 13) );
  	edtCidade = new wx::TextCtrl(this, (int)idx::ID_FRMCADENDERECOS_EDTCIDADE, 
      "", System::Drawing::Point(88, 104), System::Drawing::Size(184, 21) );
    edtCidade->Enabled = false;
  	labEstado = new wx::StaticText(this, (int)idx::ID_FRMCADENDERECOS_LABESTADO, 
      "Estado: ", System::Drawing::Point(280, 88), System::Drawing::Size(56, 13) );
    AdicionaCombo();
  	cmbEstado = new wx::Choice(this, (int)idx::ID_FRMCADENDERECOS_CMBESTADO, 
      System::Drawing::Point(280, 104), System::Drawing::Size(144, 21), vsEstados );
    cmbEstado->Enabled = false;

    IniciaBotoes();
  
    btnPesquisar->SetSize(440, 8, 75, 25);
    btnAnterior->SetSize(440, 40, 75, 25);
    btnProximo->SetSize(440, 72, 75, 25);
    btnSalvar->SetSize(200, 136, 75, 25);
    btnExcluir->SetSize(280, 136, 75, 25);
    btnLimpar->SetSize(360, 136, 75, 25);
    btnFechar->SetSize(440, 136, 75, 25);
  
    //FazLayout();
  
    edtDescricao->MaxLength = 40;
    edtBairro->MaxLength = 15;
    edtCEP->MaxLength = 8;
    edtCidade->MaxLength = 30;

    EVT_BUTTON((int)idt::ID_FRMCADASTROS_BTNNOVO, 
      new wx::EventListener( this, OnBtnNovo ));
    EVT_BUTTON((int)idt::ID_FRMCADASTROS_BTNPESQUISAR, 
      new wx::EventListener( this, OnBtnPesquisar ));
    EVT_BUTTON((int)idt::ID_FRMCADASTROS_BTNANTERIOR, 
      new wx::EventListener( this, OnBtnAnterior ));
    EVT_BUTTON((int)idt::ID_FRMCADASTROS_BTNPROXIMO, 
      new wx::EventListener( this, OnBtnProximo ));
    EVT_BUTTON((int)idt::ID_FRMCADASTROS_BTNSALVAR, 
      new wx::EventListener( this, OnBtnSalvar ));
    EVT_BUTTON((int)idt::ID_FRMCADASTROS_BTNEXCLUIR, 
      new wx::EventListener( this, OnBtnExcluir ));
    EVT_BUTTON((int)idt::ID_FRMCADASTROS_BTNLIMPAR, 
      new wx::EventListener( this, OnBtnLimpar ));
    EVT_BUTTON((int)idt::ID_FRMCADASTROS_BTNFECHAR, 
      new wx::EventListener( this, OnBtnFechar ));
  	EVT_CLOSE( new wx::EventListener( this, OnClose ) );
  }

  Void frmCadEnderecos::FazLayout(){
    wx::BoxSizer* sizer_1 = new wx::BoxSizer(wx::Orientation::wxVERTICAL);
    wx::BoxSizer* sizer_2 = new wx::BoxSizer(wx::Orientation::wxHORIZONTAL);
    wx::BoxSizer* sizer_3 = new wx::BoxSizer(wx::Orientation::wxHORIZONTAL);
    wx::BoxSizer* sizer_4 = new wx::BoxSizer(wx::Orientation::wxVERTICAL);
    wx::BoxSizer* sizer_5 = new wx::BoxSizer(wx::Orientation::wxVERTICAL);
    wx::BoxSizer* sizer_6 = new wx::BoxSizer(wx::Orientation::wxHORIZONTAL);
    wx::GridSizer* grid_sizer_6 = new wx::GridSizer(2, 1, 0, 0);
    wx::GridSizer* grid_sizer_5 = new wx::GridSizer(2, 1, 0, 0);
    wx::GridSizer* grid_sizer_4 = new wx::GridSizer(2, 1, 0, 0);
    wx::BoxSizer* sizer_8 = new wx::BoxSizer(wx::Orientation::wxHORIZONTAL);
    wx::GridSizer* grid_sizer_9 = new wx::GridSizer(2, 1, 0, 0);
    wx::GridSizer* grid_sizer_8 = new wx::GridSizer(2, 1, 0, 0);
    wx::BoxSizer* sizer_7 = new wx::BoxSizer(wx::Orientation::wxHORIZONTAL);
    wx::GridSizer* grid_sizer_7 = new wx::GridSizer(2, 1, 0, 0);
    grid_sizer_7->Add(labCodigo, 0, wx::Alignment::wxALIGN_BOTTOM, 0);
    grid_sizer_7->Add(edtCodigo, 0, 0, 0);
    sizer_7->Add(grid_sizer_7, 0, 0, 0);
    sizer_7->Add(btnNovo, 0, wx::Direction::wxLEFT|wx::Alignment::wxALIGN_BOTTOM, 5);
    sizer_5->Add(sizer_7, 1, 0, 0);
    grid_sizer_8->Add(labDescricao, 0, wx::Alignment::wxALIGN_BOTTOM, 0);
    grid_sizer_8->Add(edtDescricao, 0, 0, 0);
    sizer_8->Add(grid_sizer_8, 0, 0, 0);
    grid_sizer_9->Add(labBairro, 0, wx::Alignment::wxALIGN_BOTTOM, 0);
    grid_sizer_9->Add(edtBairro, 0, 0, 0);
    sizer_8->Add(grid_sizer_9, 0, wx::Direction::wxLEFT, 2);
    sizer_5->Add(sizer_8, 1, wx::Stretch::wxEXPAND, 0);
    grid_sizer_4->Add(labCEP, 0, wx::Alignment::wxALIGN_BOTTOM, 0);
    grid_sizer_4->Add(edtCEP, 0, 0, 0);
    sizer_6->Add(grid_sizer_4, 0, 0, 0);
    grid_sizer_5->Add(labCidade, 0, wx::Alignment::wxALIGN_BOTTOM, 0);
    grid_sizer_5->Add(edtCidade, 0, 0, 0);
    sizer_6->Add(grid_sizer_5, 0, wx::Direction::wxLEFT, 4);
    grid_sizer_6->Add(labEstado, 0, wx::Alignment::wxALIGN_BOTTOM, 0);
    grid_sizer_6->Add(cmbEstado, 0, 0, 0);
    sizer_6->Add(grid_sizer_6, 0, wx::Direction::wxLEFT, 5);
    sizer_5->Add(sizer_6, 0, 0, 0);
    sizer_3->Add(sizer_5, 0, 0, 0);
    sizer_4->Add(btnPesquisar, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT, 2);
    sizer_4->Add(btnAnterior, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT, 2);
    sizer_4->Add(btnProximo, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT, 2);
    sizer_3->Add(sizer_4, 1, 0, 0);
    sizer_1->Add(sizer_3, 1, wx::Direction::wxALL|wx::Stretch::wxEXPAND, 4);
    sizer_2->Add(btnSalvar, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 2);
    sizer_2->Add(btnExcluir, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 2);
    sizer_2->Add(btnLimpar, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 2);
    sizer_2->Add(btnFechar, 0, wx::Direction::wxALL|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 2);
    sizer_1->Add(sizer_2, 0, wx::Direction::wxRIGHT|wx::Direction::wxBOTTOM|
      wx::Alignment::wxALIGN_RIGHT|wx::Alignment::wxALIGN_BOTTOM, 4);
    AutoLayout = true;
    SetSizer(sizer_1);
    sizer_1->Fit(this);
    sizer_1->SetSizeHints(this);
  }

	Boolean frmCadEnderecos::ValidaDados(Boolean bTodosDados){
    if (RotinasGlobais::Rotinas->ValidaCampo(edtCodigo->Value, 
      Resource::STR_CODNINF))
      return false;
    if (bTodosDados) {
      if (RotinasGlobais::Rotinas->ValidaCampo(edtDescricao->Value, 
        Resource::STR_ENDNINF))
      	return false;
      if (RotinasGlobais::Rotinas->ValidaCampo(edtBairro->Value, 
        Resource::STR_BAIRRONINF))
      	return false;
      if (RotinasGlobais::Rotinas->ValidaCampo(edtCEP->Value, 
        Resource::STR_CEPNINF))
      	return false;
      if (RotinasGlobais::Rotinas->ValidaCampo(edtCidade->Value, 
        Resource::STR_CIDADENINF))
      	return false;
    }
		return true;
	}

  Void frmCadEnderecos::OnClose( System::Object* sender, wx::Event* e ){
   this->Parent->Close();
  	e->Skip();
  }
  
  Void frmCadEnderecos::OnBtnNovo( System::Object* sender, wx::Event* e ){
    btnNovoClick();
    InformaLimpaDados(false);
    HabilitaDados(true);
    NovoDado(ConsultasSQL::ConsSQL->Endereco('N', edtCodigo->Value,
      "","","","",""));	
    e->Skip();
  }
  
  Void frmCadEnderecos::OnBtnPesquisar( System::Object* sender, wx::Event* e ){
    if (PesquisarDados(ConsultasSQL::ConsSQL->Endereco('S', edtCodigo->Value,
      "","","","",""), Resource::STR_ENDNENC)) {
      InformaLimpaDados(true);
      HabilitaDados(true);
    }
    e->Skip();
  }
  
  Void frmCadEnderecos::OnBtnAnterior( System::Object* sender, wx::Event* e ){
    if (registroAnterior())
      InformaLimpaDados(true);
    e->Skip();
  }
  
  Void frmCadEnderecos::OnBtnProximo( System::Object* sender, wx::Event* e ){
    if (registroProximo())
      InformaLimpaDados(true);
    e->Skip();
  }
  
  Void frmCadEnderecos::OnBtnSalvar( System::Object* sender, wx::Event* e ){
    if (SalvarDados(ValidaDados(true),
      ConsultasSQL::ConsSQL->Endereco('S', edtCodigo->Value,
        "","","","",""),
      ConsultasSQL::ConsSQL->Endereco('U', edtCodigo->Value, 
        edtDescricao->Value,edtBairro->Value,edtCEP->Value,
        edtCidade->Value,cmbEstado->StringSelection),
      ConsultasSQL::ConsSQL->Endereco('I', edtCodigo->Value, 
        edtDescricao->Value,edtBairro->Value,edtCEP->Value,
        edtCidade->Value,cmbEstado->StringSelection), true)) {
      InformaLimpaDados(false);
      HabilitaDados(false);
    }
    e->Skip();
  }
  
  Void frmCadEnderecos::OnBtnExcluir( System::Object* sender, wx::Event* e ){
    if (ExcluirDados(ValidaDados(false),
      ConsultasSQL::ConsSQL->Endereco('D', edtCodigo->Value,
      "","","","",""))) {
      InformaLimpaDados(false);
      HabilitaDados(false);
    }
    e->Skip();
  }
  
  Void frmCadEnderecos::OnBtnFechar( System::Object* sender, wx::Event* e ){
    this->Close(true);
    e->Skip();
  }
  
  Void frmCadEnderecos::OnBtnLimpar( System::Object* sender, wx::Event* e ){
    btnLimparClick();
    InformaLimpaDados(false);
    HabilitaDados(false);
    e->Skip();
  }

  Void frmCadEnderecos::InformaLimpaDados(Boolean bInformar){
    if (bInformar) {
      edtBairro->Value = Campos->Item[2]->ToString();
      edtCEP->Value = Campos->Item[3]->ToString();
      edtCidade->Value = Campos->Item[4]->ToString();
      cmbEstado->StringSelection = Campos->Item[5]->ToString();
    } else {
      edtBairro->Clear();
      edtCEP->Clear();
      edtCidade->Clear();
      cmbEstado->Selection = 0;
    }
  }
  
  Void frmCadEnderecos::HabilitaDados(Boolean bHabilita){
    edtBairro->Enabled = bHabilita;
    edtCEP->Enabled = bHabilita;  
    edtCidade->Enabled = bHabilita;
    cmbEstado->Enabled = bHabilita;  
  }

  Void frmCadEnderecos::AdicionaCombo(){
    vsEstados = new String*[26];
    vsEstados->SetValue(Resource::STR_PE, 0);
    vsEstados->SetValue(Resource::STR_AC, 1);
    vsEstados->SetValue(Resource::STR_AL, 2);
    vsEstados->SetValue(Resource::STR_AP, 3);
    vsEstados->SetValue(Resource::STR_AM, 4);
    vsEstados->SetValue(Resource::STR_BA, 5);
    vsEstados->SetValue(Resource::STR_BR, 6);
    vsEstados->SetValue(Resource::STR_CE, 7);
    vsEstados->SetValue(Resource::STR_DF, 8);
    vsEstados->SetValue(Resource::STR_ES, 9);
    vsEstados->SetValue(Resource::STR_GO, 10);
    vsEstados->SetValue(Resource::STR_MT, 11);
    vsEstados->SetValue(Resource::STR_MS, 12);
    vsEstados->SetValue(Resource::STR_MG, 13);
    vsEstados->SetValue(Resource::STR_PA, 14);
    vsEstados->SetValue(Resource::STR_PB, 15);
    vsEstados->SetValue(Resource::STR_PR, 16);
    vsEstados->SetValue(Resource::STR_SC, 17);
    vsEstados->SetValue(Resource::STR_RN, 18);
    vsEstados->SetValue(Resource::STR_RS, 19);
    vsEstados->SetValue(Resource::STR_RJ, 20);
    vsEstados->SetValue(Resource::STR_RO, 21);
    vsEstados->SetValue(Resource::STR_RR, 22);
    vsEstados->SetValue(Resource::STR_SP, 23);
    vsEstados->SetValue(Resource::STR_SE, 24);
    vsEstados->SetValue(Resource::STR_TO, 25);
  }

}
